internal/controller/controllers/preprovisioningimage_controller.go

```diff
@@ -396,9 +396,133 @@ func (r *PreprovisioningImageReconciler) mapInfraEnvPPI() func(ctx context.Conte
+		if err != nil {
+			log.WithError(err).Warningf("Failed to get release architecture for (%s)", cv.Status.Desired.Image)
+			return nil
+		}
+		image, err = r.OcRelease.GetIronicAgentImage(log, cv.Status.Desired.Image, r.ReleaseImageMirror, infraEnvInternal.PullSecret)
+		if err != nil {
+			log.WithError(err).Warningf("Failed to get ironic agent image from release (%s)", cv.Status.Desired.Image)
+			return nil
+		}
+		r.hubIronicAgentImage = image
+		r.hubReleaseArchitectures = architectures
+		log.Infof("Caching hub ironic agent image %s for architectures %v", image, architectures)
+	}
+
+	if !funk.Contains(architectures, infraEnvInternal.CPUArchitecture) {
+		log.Warningf("Release image architectures %v do not match infraEnv architecture %s", architectures, infraEnvInternal.CPUArchitecture)
+		return nil
+	}
+
+	log.Infof("Setting ironic agent image (%s) from HUB cluster", image)
+	return &ICCConfig{
+		IronicAgentImage: image,
+	}
+}
+
+func (r *PreprovisioningImageReconciler) getIronicDefaultConfig(log logrus.FieldLogger, infraEnvInternal *common.InfraEnv) *ICCConfig {
+	var ironicAgentImage string
+	if infraEnvInternal.CPUArchitecture == common.ARM64CPUArchitecture {
+		ironicAgentImage = r.Config.BaremetalIronicAgentImageForArm
+	} else {
+		ironicAgentImage = r.Config.BaremetalIronicAgentImage
+	}
+
+	log.Infof("Setting default ironic agent image (%s)", ironicAgentImage)
+	return &ICCConfig{
+		IronicAgentImage: ironicAgentImage,
+	}
+}
+
+func (r *PreprovisioningImageReconciler) getIronicConfig(ctx context.Context, log logrus.FieldLogger, infraEnv *aiv1beta1.InfraEnv, infraEnvInternal *common.InfraEnv) (*ICCConfig, error) {
+	var iccConfig *ICCConfig
+
+	iccConfig = r.getIronicConfigFromUserOverride(log, infraEnv)
+
+	if iccConfig == nil {
+		iccConfig = r.getIronicConfigFromBMOConfig(ctx, log, infraEnvInternal)
+	}
+
+	if iccConfig == nil {
+		iccConfig = r.getIronicConfigFromHUB(ctx, log, infraEnvInternal)
+	}
+
+	if iccConfig == nil {
+		iccConfig = r.getIronicDefaultConfig(log, infraEnvInternal)
+	}
+
+	if iccConfig == nil {
+		return nil, fmt.Errorf("Failed to determine ironic config")
+	}
+
+	if iccConfig.IronicBaseURL == "" && iccConfig.IronicInspectorBaseUrl == "" {
+		err := r.fillIronicServiceURLs(ctx, infraEnvInternal, iccConfig)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	log.Infof("Ironic Agent Image is (%s) Ironic URL is (%s) Inspector URL is (%s)",
+		iccConfig.IronicAgentImage,
+		iccConfig.IronicBaseURL,
+		iccConfig.IronicInspectorBaseUrl)
+	return iccConfig, nil
+}
+
 // AddIronicAgentToInfraEnv updates the infra-env in the database with the ironic agent ignition config if required
 // returns true when the infra-env was updated, false otherwise
 func (r *PreprovisioningImageReconciler) AddIronicAgentToInfraEnv(ctx context.Context, log logrus.FieldLogger, infraEnv *aiv1beta1.InfraEnv) (bool, error) {
+	log = log.WithFields(logrus.Fields{
```

Comment:
Should we maybe do this earlier for all functions as soon as we know the infraenv?

Reply:
changed it!